runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -157,112 +177,232 @@ class SharedLibraryManager final {
+    [[nodiscard]] auto getTapeRecorderStatus() const -> bool { return tape_recording; }
 
-    [[nodiscard]] QuantumDevice *loadDevice(std::string filename)
+    [[nodiscard]] auto getDeviceInfo() const -> std::tuple<std::string, std::string, std::string>
     {
-        _device_so_ptr = std::make_unique<SharedLibraryManager>(filename);
-        std::string factory_name{_device_name + "Factory"};
-        void *f_ptr = _device_so_ptr->getSymbol(factory_name);
-        return f_ptr ? reinterpret_cast<decltype(GenericDeviceFactory) *>(f_ptr)(
-                           _device_kwargs.c_str())
-                     : nullptr;
+        return {rtd_lib, rtd_name, rtd_kwargs};
     }
 
-    [[nodiscard]] bool initDevice(std::string_view rtd_lib, std::string_view rtd_name,
-                                  std::string_view rtd_kwargs)
+    [[nodiscard]] auto getDeviceName() const -> const std::string & { return rtd_name; }
+
+    friend std::ostream &operator<<(std::ostream &os, const RTDeviceType &device)
     {
-        // Reset the device pointer
-        _device_ptr.reset(nullptr);
+        os << "RTD, name: " << device.rtd_name << " lib: " << device.rtd_lib
+           << " kwargs: " << device.rtd_kwargs;
+        return os;
+    }
+};
 
-        // Set the device info.
-        _device_name = rtd_name;
-        _device_kwargs = rtd_kwargs;
+class ExecutionContext final {
+  private:
+    // Device pool
+    std::unordered_map<size_t, std::pair<RTDeviceStatus, std::shared_ptr<RTDeviceType>>>
+        device_pool;
+    size_t pool_counter{0}; // Counter for generating unique keys for devices
 
-        // Convert to a mutable string
-        std::string rtd_lib_str{rtd_lib};
+    bool initial_tape_recorder_status;
 
-        // LCOV_EXCL_START
-        // The following if-elif is required for C++ tests where these backend devices
-        // are linked in the interface library of the runtime. (check runtime/CMakeLists.txt)
-        // Besides, this provides support for runtime device (RTD) libraries added to the system
-        // path. This maintains backward compatibility for specifying a device using its name.
-        // TODO: This support may need to be removed after updating the C++ unit tests.
-        if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
-            _device_name =
-                (rtd_lib == "lightning.qubit") ? "LightningSimulator" : "LightningKokkosSimulator";
-            update_device_pl2runtime(rtd_lib_str, "lightning");
-        }
-        else if (rtd_lib == "braket.aws.qubit" || rtd_lib == "braket.local.qubit") {
-            _device_name = "OpenQasmDevice";
-            update_device_pl2runtime(rtd_lib_str, "openqasm");
-        }
-        // LCOV_EXCL_STOP
+    // ExecutionContext pointers
+    std::unique_ptr<MemoryManager> memory_man_ptr{nullptr};
+    std::unique_ptr<PythonInterpreterGuard> py_guard{nullptr};
```

Comment:
This guard is used to avoid double initialization of the Python interpreter for Braket devices. If you call runtime from the frontend, we don't need to re-initialize the interpreter (using CPython functions). Otherwise,  we initialize a new Python interpreter instance in the runtime and keep it until the end of the program scope. This guard takes care of the lifetime of this interpreter instance.

Reply:
Thanks!